lib/app/modules/us_core_r4/clinicalnotes_sequence.rb

```diff
@@ -0,0 +1,228 @@
+        returned will be checked for consistency against the [US Core Clinical Notes Guidance](https://build.fhir.org/ig/HL7/US-Core-R4/clinical-notes-guidance.html)
+
+      )
+
+      @clinicalnotes_found = false
+
+      def test_clinicalnotes_documentreference(type_code)
+        skip 'No Clinical Notes appear to be available for this patient. Please use patients with more information.' unless @clinicalnotes_found
+
+        assert @must_have_type_code[type_code], "Clinical Notes shall have at least one DocumentReference with type #{type_code}"
+      end
+
+      def test_clinicalnotes_diagnosticreport(category_code)
+        skip 'No Clinical Notes appear to be available for this patient. Please use patients with more information.' unless @clinicalnotes_found
+
+        patient_val = @instance.patient_id
+        search_params = { 'patient': patient_val, 'category': category_code }
+
+        reply = get_resource_by_params(versioned_resource_class('DiagnosticReport'), search_params)
+        assert_response_ok(reply)
+        assert_bundle_response(reply)
+
+        resource_count = reply&.resource&.entry&.length || 0
+        @resources_found = true if resource_count.positive?
+
+        assert @resources_found, "Clinical Notes shall have at least one DiagnosticReport with category #{category_code}"
+
+        diagnosticreport_ary = reply&.resource&.entry&.map { |entry| entry&.resource }
+
+        diagnosticreport_ary&.each do |resource|
+          id = resource&.id
+
+          resource&.presentedForm&.each do |attachment|
+            url_value = attachment&.url
+            next if url_value.nil?
+
+            assert @attachment_url.key?(url_value), "Attachment #{url_value} referenced in DiagnosticReport/#{id} but not in any DocumentReference"
+            @attachment_url[url_value][:flag] = true
+          end
+        end
+      end
+
+      test 'Server returns expected results from DocumentReference search by patient+clinicalnotes' do
+        metadata do
+          id '01'
+          link 'https://build.fhir.org/ig/HL7/US-Core-R4/clinical-notes-guidance.html'
+          desc %(
+          )
+          versions :r4
+        end
+
+        patient_val = @instance.patient_id
+        category_val = 'clinical-note'
+        search_params = { 'patient': patient_val, 'category': category_val }
+        @attachment_url = {}
+
+        reply = get_resource_by_params(versioned_resource_class('DocumentReference'), search_params)
+        assert_response_ok(reply)
+        assert_bundle_response(reply)
+
+        resource_count = reply&.resource&.entry&.length || 0
+        @clinicalnotes_found = true if resource_count.positive?
+
+        skip 'No Clinical Notes appear to be available for this patient. Please use patients with more information.' unless @clinicalnotes_found
+
+        documentreference_ary = reply&.resource&.entry&.map { |entry| entry&.resource }
+
+        @must_have_type_code = {
```

Comment:
I don't think a `Hash` is the right data type here. If all we want to do is check inclusion in a group, I think a `Set` would be more appropriate. We have a set of the codes we expect to see, and a set of codes we have seen. We compare the two sets to determine whether all codes have been seen.

Reply:
fixed